src/rust/engine/fs/src/lib.rs

```diff
@@ -455,14 +469,77 @@ impl PosixFS {
       .to_boxed()
   }
 
+  ///
+  /// Makes a Stat for path_for_stat relative to absolute_path_to_root.
+  ///
+  fn stat_internal<F>(
+    path_for_stat: PathBuf,
+    file_type: std::fs::FileType,
+    absolute_path_to_root: &Path,
+    get_metadata: F,
+  ) -> Result<Stat, io::Error>
+  where
+    F: FnOnce() -> Result<fs::Metadata, io::Error>,
+  {
+    if !path_for_stat.is_relative() {
+      return Err(io::Error::new(
+        io::ErrorKind::InvalidInput,
+        format!(
+          "Argument path_for_stat to PosixFS::stat must be relative path, got {:?}",
+          path_for_stat
+        ),
+      ));
+    }
+    if !absolute_path_to_root.is_absolute() {
+      return Err(io::Error::new(
+        io::ErrorKind::InvalidInput,
+        format!(
+          "Argument absolute_path_to_root to PosixFS::stat must be absolute path, got {:?}",
+          absolute_path_to_root
+        ),
+      ));
+    }
+    if file_type.is_dir() {
+      Ok(Stat::Dir(Dir(path_for_stat)))
+    } else if file_type.is_file() {
+      let is_executable = get_metadata()?.permissions().mode() & 0o100 == 0o100;
+      Ok(Stat::File(File {
+        path: path_for_stat,
+        is_executable: is_executable,
+      }))
+    } else if file_type.is_symlink() {
+      Ok(Stat::Link(Link(path_for_stat)))
+    } else {
+      Err(io::Error::new(
+        io::ErrorKind::InvalidData,
+        format!(
+          "Expected File, Dir or Link, but {:?} (relative to {:?}) was a {:?}",
+          path_for_stat,
+          absolute_path_to_root,
+          file_type
+        ),
+      ))
+    }
+  }
+
+  pub fn stat(&self, relative_path: PathBuf) -> Result<Stat, io::Error> {
+    let metadata = fs::symlink_metadata(self.root.0.join(&relative_path))?;
```

Comment:
Calling this ahead of time sortof defeats the purpose of making it lazy.

Maybe you could change the function to:
```
F: FnOnce(&Path) -> Result<fs::Metadata, io::Error>
```
The `stat` version would actually use the borrowed path to make the syscall, while the `read_dir` version would ignore it.

Reply:
For the `stat` call, we always need to proactively stat the file because we need to find out the `file_type`. (Note that we pass `metadata.file_type()` as an arg to `stat_internal`).

For the `scandir` call, we already have the `file_type` from the `DirEntry`, so we only need to stat the file in the case that it's a File, to get permissions.

I made this lazy so that the `scandir` case can make the stat call optional as needed, and the `stat` case can re-use the output of the stat it already performed to find out the `file_type`.

Would renaming anything help here? Or adding a comment somewhere? Or am I missing something?

Reply:
Ah, no: makes sense.